xio-core/src/main/java/com/xjeffrose/xio/SSL/TlsConfig.java

```diff
@@ -172,33 +168,124 @@ private static ApplicationProtocolConfig buildAlpnConfig(Config config) {
 
-  public static TlsConfig fromConfig(String key, Config config) {
-    return new TlsConfig(config.getConfig(key));
+  /**
+   * Creates a new builder with values from the given config object.
+   *
+   * @throws ConfigException if any of the keys present in the Config object have unexpected value
+   *     types
+   */
+  public static Builder builderFrom(Config config) {
+    Builder builder = TlsConfig.builder();
+    getOptionalPath(config, "useSsl", Config::getBoolean).map(builder::useSsl);
+    getOptionalPath(config, "logInsecureConfig", Config::getBoolean)
+        .map(builder::logInsecureConfig);
+
+    getOptionalPath(config, "x509TrustedCertPaths", Config::getStringList)
+        .map(TlsConfig::buildCerts)
+        .map(builder::trustedCerts);
+
+    getOptionalPath(config, "privateKeyPath", Config::getString)
+        .map(TlsConfig::readPathFromValue)
+        .map(TlsConfig::parsePrivateKeyFromPem)
+        .map(builder::privateKey);
+
+    getOptionalPath(config, "x509CertPath", Config::getString)
+        .map(TlsConfig::readPathFromValue)
+        .map(TlsConfig::parseX509CertificateFromPem)
+        .map(builder::certificate);
+
+    getOptionalPath(config, "x509CertChainPaths", Config::getStringList)
+        .map(TlsConfig::buildCerts)
+        .map(builder::certChain);
+
+    getOptionalPath(config, "useOpenSsl", Config::getBoolean).map(builder::useOpenSsl);
+
+    getOptionalPath(config, "alpn", Config::getConfig)
+        .map(TlsConfig::buildAlpnConfig)
+        .map(builder::alpnConfig);
+
+    getOptionalPath(config, "ciphers", Config::getStringList).map(builder::ciphers);
+    getOptionalPath(
+            config, "clientAuth", (configArg, key) -> configArg.getEnum(ClientAuth.class, key))
+        .map(builder::clientAuth);
+    getOptionalPath(config, "enableOcsp", Config::getBoolean).map(builder::enableOcsp);
+    getOptionalPath(config, "protocols", Config::getStringList).map(builder::protocols);
+    getOptionalPath(config, "sessionCacheSize", Config::getLong).map(builder::sessionCacheSize);
+    getOptionalPath(config, "sessionTimeout", Config::getLong).map(builder::sessionTimeout);
+
+    return builder;
   }
 
-  public static TlsConfig fromConfig(String key) {
-    return fromConfig(key, ConfigFactory.load());
+  @lombok.Builder(builderClassName = "Builder")
+  private TlsConfig(
+      Boolean useSsl,
+      Boolean useOpenSsl,
+      Boolean logInsecureConfig,
+      PrivateKey privateKey,
+      X509Certificate certificate,
+      List<X509Certificate> certChain,
+      List<X509Certificate> trustedCerts,
+      ApplicationProtocolConfig alpnConfig,
+      List<String> ciphers,
+      ClientAuth clientAuth,
+      Boolean enableOcsp,
+      List<String> protocols,
+      Long sessionCacheSize,
+      Long sessionTimeout) {
+    // Validate that *all* parameters were set in the builder.
+    String errorMessage = "{} must be set in builder";
+    Preconditions.checkNotNull(useSsl, errorMessage, "useSsl");
```

Comment:
i'd expect lombok to be able to make this not null requirement check with an annotation

Reply:
`lombok.Builder` is lacking in generated validation hooks, for that reason we've moved this sort of functionality over to google auto builder, we can refactor later.